realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -2474,3 +2486,480 @@ void TemperatureDiagnostics::diagnostics(diagnostic_updater::DiagnosticStatusWra
+
+struct Clock
+{
+    using InternalClock = std::chrono::steady_clock;
+
+    Clock() : start(InternalClock::now()) {}
+    void restart() {
+        start = InternalClock::now();
+    }
+
+    double getElapsedSecs() {
+        return static_cast<std::chrono::duration<double>>(InternalClock::now() - start).count();
+    }
+
+private:
+    InternalClock::time_point start;
+};
+
+void BaseRealSenseNode::nomagicSetupService(stream_index_pair stream, bool is_aligned_depth)
+{
+    // name + [index]: color, depth, infra1, infra2, etc.
+    std::string stream_name = (is_aligned_depth ? "aligned_depth_to_" : "") + STREAM_NAME(stream);
+    std::string full_service_name = stream_name + "/get_latest_frame";
+
+    auto& service_container = is_aligned_depth
+                            ? nomagic_get_latest_aligned_frame_servers
+                            : nomagic_get_latest_frame_servers;
+
+    // Trampoline to pass stream parameter to the callback.
+    boost::function<bool(GetLatestFrame::Request&, GetLatestFrame::Response&)> callback =
+            [=](GetLatestFrame::Request& request, GetLatestFrame::Response& response) {
+        return nomagicGetLatestFrameCallback(stream, is_aligned_depth, request, response);
+    };
+
+    service_container.insert({stream, _node_handle.advertiseService(full_service_name, callback)});
+
+    ROS_INFO_STREAM("[NOMAGIC] Successfully started service " << full_service_name);
+}
+
+void BaseRealSenseNode::nomagicSetup()
+{
+    // The minimal number of frames to make the temporal filter work reasonably == 9
+    // where reasonable == having persistence settings working as expected
+    // and reasonable != equivalent to filtering (and thus accumulating) every frame.
+    // Details: librealsense/src/proc/temporal-filter.cpp
+    nomagic_frameset_queue.resize(nomagic_lazy_filtering ? nomagic_lazy_filtering_frame_history_size : 1);
+
+    for (auto& stream : IMAGE_STREAMS) {
+        if (!_enable[stream]) {
+            continue;
+        }
+
+        nomagic_expected_streams.insert(stream);
+        nomagicSetupService(stream, false);
+
+        bool advertise_aligned_depth = (_align_depth && (stream != DEPTH) && (stream != CONFIDENCE) && stream.second < 2);
+        if (advertise_aligned_depth) {
+            nomagicSetupService(stream, true);
+        }
+    }
+
+    // Setup diagnostics
+    nomagic_frameset_fragmentation_diagnostics.add("[NOMAGIC] Framesets Fragmentation Status",
+                                                   this, &BaseRealSenseNode::nomagicFramesetsDiagnosticsCallback);
+    nomagic_frameset_fragmentation_diagnostics.setHardwareID(_serial_no);
+
+    // Setup muxer
+    nomagic_muxer.start([&](rs2::frame frame) {
+        nomagicStoreFramesetForLazyProcessing(frame.as<rs2::frameset>());
+    });
+}
+
+void BaseRealSenseNode::nomagicMuxerCallback(rs2::frame frame, rs2::frame_source& src)
+{
+    auto frameset = frame.as<rs2::frameset>();
```

Comment:
Opt: it seems this could be extracted into some "FrameAccumulator" class that can be independent of the rest of the node.